### PR TITLE
Check list length and switch expressions in serialize functions

### PR DESCRIPTION
**Pull request by eduardosm**
Fixes https://github.com/psychon/x11rb/issues/355

### CODE REVIEW

File: src/protocol/xkb.rs [hunk 1]
```diff
@@ -8336,16 +8347,17 @@ pub struct SetNamesAuxBitcase8 {
...
-    fn serialize(&self) -> Self::Bytes {
+impl SetNamesAuxBitcase8 {
+    fn serialize(&self, n_types: u8) -> Vec<u8> {
         let mut result = Vec::new();
-        self.serialize_into(&mut result);
+        self.serialize_into(&mut result, n_types);
         result
     }
-    fn serialize_into(&self, bytes: &mut Vec<u8>) {
+    fn serialize_into(&self, bytes: &mut Vec<u8>, n_types: u8) {
```
Comment (eduardosm) [inline src/protocol/xkb.rs:8356]: What do you think about this? Structures that have external parsing parameters will also receive those parameters in their serialize function in order to check for consistency. Those structs will not implement `Serialize` any longer.

Comment (psychon) [reply src/protocol/xkb.rs:8356]: From the example you picked, this seems to just "shuffle" some code around, i.e. move the `assert` from the caller to the callee.
However, from other code in the diff, it seems like this actually makes some `assert`s possible that were not possible before. I am all for defensively using asserts to ensure stuff is correct, so :+1:

**Comment by eduardosm** [comment]
I think this is ready. I added `#[allow(dead_code, unused_variables)]` because functions and arguments are not always used.

File: src/protocol/render.rs [hunk 2]
```diff
@@ -2263,14 +2263,14 @@ pub struct ChangePictureAux {
...
+impl ChangePictureAux {
+    fn serialize(&self, value_mask: u32) -> Vec<u8> {
         let mut result = Vec::new();
-        self.serialize_into(&mut result);
+        self.serialize_into(&mut result, value_mask);
         result
     }
-    fn serialize_into(&self, bytes: &mut Vec<u8>) {
+    fn serialize_into(&self, bytes: &mut Vec<u8>, value_mask: u32) {
         if let Some(repeat) = self.repeat {
```
Comment (psychon) [inline src/protocol/render.rs:2274]: Would it make sense to `assert_eq!(value_mask, self.value_mask());`? It feels like that would get rid of most of the unused parameters.

**Pull request merged**